include/revng/AutoEnforcer/AutoEnforcer.h

```diff
@@ -0,0 +1,183 @@
+  using const_iterator = std::vector<Step>::const_iterator;
+  using InvalidationMap = llvm::StringMap<BackingContainersStatus>;
+
+  explicit PipelineRunner(Context &C) : AEContext(&C) {}
+
+  template<typename... EnforcerWrappers>
+  void addStep(std::string StepName, EnforcerWrappers &&...EnfWrappers) {
+    CommittedRegistry = true;
+    Pipeline.add(Step(std::move(StepName),
+                      Registry.createEmpty(),
+                      std::forward<EnforcerWrappers>(EnfWrappers)...));
+  }
+
+  llvm::Error deduceInvalidations(InvalidationMap &Invalidations) const {
+    return Pipeline.getInvalidations(Invalidations);
+  }
+
+  const BackingContainers &getStartingContainers() const {
+    return Pipeline.getStartingContainers();
+  }
+
+  BackingContainers &getStartingContainers() {
+    return Pipeline.getStartingContainers();
+  }
+
+  template<typename BackingContainerType>
+  const BackingContainerType &getStartingContainer(llvm::StringRef Name) const {
+    return Pipeline.getStartingContainer<BackingContainerType>(Name);
+  }
+
+  template<typename BackingContainerType>
+  BackingContainerType &getOrCreateStartingContainer(llvm::StringRef Name) {
+    return Pipeline.getOrCreateStartingContainer<BackingContainerType>(Name);
+  }
+
+  bool isStartingContainerEmpty(llvm::StringRef Name) const {
+    return Pipeline.getStartingContainers().isContainerEmpty(Name);
+  }
+
+  const BackingContainers &getFinalContainers() const {
+    return Pipeline.getFinalContainers();
+  }
+  BackingContainers &getFinalContainers() {
+    return Pipeline.getFinalContainers();
+  }
+
+  template<typename BackingContainerType>
+  const BackingContainerType &getFinalContainer(llvm::StringRef Name) const {
+    return Pipeline.getFinalContainer<BackingContainerType>(Name);
+  }
+
+  template<typename BackingContainerType>
+  BackingContainerType &getOrCreateFinalContainer(llvm::StringRef Name) {
+    return Pipeline.getOrCreateFinalContainer<BackingContainerType>(Name);
+  }
+
+  llvm::Error
+  run(const BackingContainersStatus &Targets, llvm::raw_ostream *OS = nullptr) {
+    return Pipeline.run(*AEContext, Targets, OS);
+  }
+
+  template<typename BackingContainerType, typename... Args>
+  void registerContainerFactory(llvm::StringRef Name, Args &&...Arguments) {
+    revng_assert(not CommittedRegistry);
+    using BCT = BackingContainerType;
+    Registry.registerContainerFactory<BCT>(Name,
+                                           std::forward<Args>(Arguments)...);
+  }
+
+  void
+  registerContainerFactory(llvm::StringRef Name,
+                           std::unique_ptr<BackingContainerFactory> Entry) {
+    Registry.registerContainerFactory(Name, std::move(Entry));
+  }
+
+  template<typename BackingContainer>
+  void registerDefaultConstructibleFactory(llvm::StringRef Name) {
+    revng_assert(not CommittedRegistry);
+    Registry.registerDefaultConstructibleFactory<BackingContainer>(Name);
+  }
```

Comment:
Do all of these 3 actually have a proper use case?

Reply:
yeah, the first and last are the most reasonable way one would write cpp code, the second one is used by the loader when a general way of handling the problem is needed.